_resources/port1.0/group/legacysupport-1.0.tcl

```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+#
+# This portgroup provides support for various missing library functions
+# on older OS releases.
+
+# Newest OSX release that requires legacy support.
+# Currently OSX 10.11 (Darwin 15) due to clock_gettime
+set newest_darwin_requires_legacy 15
+
+proc add_legacysupport {} {
+    
+    global prefix
+    set MPLegacyIncDir ${prefix}/include/LegacySupport
+
+    depends_lib-append port:legacy-support
+
+    # Add to configure options
+    configure.ldflags-append  -lMacportsLegacySupport
+    configure.cflags-append   -I${MPLegacyIncDir}
+    configure.cppflags-append -I${MPLegacyIncDir}
+
+    # Set env vars so gcc/clang add legacy include dir to default search paths
+    # Note using C_INCLUDE_PATH and CPLUS_INCLUDE_PATH to avoid conflicts
+    # eith MacPorts setting of CPATH
+    configure.env-append     C_INCLUDE_PATH=${MPLegacyIncDir} \
+                         CPLUS_INCLUDE_PATH=${MPLegacyIncDir}
+    build.env-append         C_INCLUDE_PATH=${MPLegacyIncDir} \
+                         CPLUS_INCLUDE_PATH=${MPLegacyIncDir}
+}
+
+if {${os.platform} eq "darwin" && ${os.major} <= ${newest_darwin_requires_legacy}} {
```

Comment:
I would prefer a solution where the port itself could specify which version requires legacy support. I'm pretty sure there will be ports which will only require this on <= 10.6, but this would unconditionally add the headers and library all the way up to 10.11, potentially even making some ports non-distributable for no reason at all.

Just curious: what will happen when we add one function to fake support for A in 10.11 (take realtime clock as a concrete example) and then another function to fake support for B in 10.13, and we'll need support for B for one port. Will the port use native support for A or the one faked by MacPorts when compiled on 10.13?

Reply:
One more thing: in case that we add another patch just for the new OS that will require an increased `newest_darwin_requires_legacy`, we would need to revbump all ports using this group even if almost none would require a change. If we always code the OS version number in need of legacy support, no revbumps would be needed. (OK, I'm just assuming that no revbumps would be needed; maybe they would still be needed just because the library would change enough to require a rebuild. But it still sounds cleaner to put this number into the port.)

We could still have this number in the PortGroup, meaning that users' request for "I want legacy support for darwin17" will be ignored on anything newer than darwin15, ...

Reply:
On the question of rev-bumping pmts when legacy-support is updated in the future, yes by the book this might be needed. In practise though I think it will not be a problem as long as these future updates only add new features, but do not change those already implemented. I have effectively, whilst developing things, been doing this and I have never seen the rev-bump check trigger an issue. So in practise I suspect this will not be a real significant problem.

Reply:
> We could still have this number in the PortGroup, meaning that users' request for "I want legacy support for darwin17" will be ignored on anything newer than darwin15, ...

I think you might have mis-understood a little how the library and headers work, most probably my fault for not explaining it well enough.

The main thing to realise is the headers and library *never* expose any new features, on any given OS, that that OS already has. If the OS already has some feature the preprocessor guards hide the implementation in the headers, and prevent the methods being built into the library.

The point of the `newest_darwin_requires_legacy` variable in the PortGroup is just an implementation detail. For OSes newer than this, neither the headers or the library add anything. The headers only `include_next` include the main system header and the library is basically empty. In this case there the PortGroup does nothing, so this variable prevents it ever being built.

On systems where the port does provide some level of support, then the library must be built with everything it provides, in order to satisfy all possible ports that use it. So individual ports 'opting in' to different levels of support is really not an option. Nor do I really see a need. If the port only needs one feature and not another, then that is all it will use.

@mojca Please let me know if I have explained all this well enough to answer your concerns.

Reply:
> > I would prefer a solution where the port itself could specify which version requires legacy support. I'm pretty sure there will be ports which will only require this on <= 10.6, but this would unconditionally add the headers and library all the way up to 10.11, potentially even making some ports non-distributable for no reason at all.
> 
> I guess I could make `newest_darwin_requires_legacy` an option that ports if they need to could change. I don't see this being required often, but perhaps useful sometimes.
> 
> Actually no, this cannot be done as only one version of the library can exist on any given system, and needs to satisfy the needs of all ports. So I think things as they are are the only way to support this. Given the fact the library by construction never implements anything the system already provides, I do not think there is a problem here.

Sorry... I was confusing even myself here... Making `newest_darwin_requires_legacy` an option each port could if they wish change is OK, as this of course does not affect how the port is built. So yes, I will add this.

Reply:
I'm sorry about the stupid question of features A and B. Yes, that was clear to me at some point (that functions would only be ifdef-ed to a specific OS version anyway), but I was confused at another point when writing that question down.

But there's still a difference of whether or not a port depends on this port, whether or not it links against the dylib (even if for no reason at all) and potentially whether or not it's distributable. I would use a different variable name though. Something like
```tcl
legacysupport.{something} 10
```
where `{something}` could be `required_by_darwin_version_upto` or ideally some better / shorter / more descriptive name. But something that starts with the portgroup's name, similar to `github.setup` etc.

Reply:
I've pushed any update that makes the newest Darwin version that support will be activated for a configurable option.